FPGA: add state_machine code sample (WIP) (#2)
haoyanwa/oneAPI-samples: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/state_machine/README.md

Reviewer: Don't forget to register your code sample in the top-level readme:

https://github.com/oneapi-src/oneAPI-samples/blob/master/DirectProgramming/C%2B%2BSYCL_FPGA/README.md

Others on the team can help you with this (Kevin, Juny, Jun Cheng, Yohann)

# `state_machine` Sample

> This project demonstrates how to make an efficient state machine design with the oneAPI FPGA compiler.

Reviewer: ```suggestion
This tutorial demonstrates how to make an efficient state machine design with the Intel® oneAPI DPC++/C++ FPGA compiler.
```

| Time to complete                  | 10 minutes

> **Note**: Even though the Intel DPC++/C++ OneAPI compiler is enough to compile for emulation, generating reports and generating RTL, there are extra software requirements for the simulation flow and FPGA compiles.
>
> To use the simulator flow, Intel® Quartus® Prime Pro Edition and one of the following simulators must be installed and accessible through your PATH:
> - Questa*-Intel® FPGA Edition
> - Questa*-Intel® FPGA Starter Edition
> - ModelSim® SE
>
> When using the hardware compile flow, Intel® Quartus® Prime Pro Edition must be installed and accessible through your PATH.
>
> :warning: Make sure you add the device files associated with the FPGA that you are targeting to your Intel® Quartus® Prime installation.

> **Note**: In oneAPI full systems, kernels that use SYCL Unified Shared Memory (USM) host allocations or USM shared allocations (and therefore the code in this tutorial) are only supported by Board Support Packages (BSPs) with USM support. Kernels that use these types of allocations can always be used to generate standalone IPs.

## Purpose

> This code sample demonstrates two different implementations of a state machine using SYCL for High-Level Synthesis (HLS) on Intel FPGAs: the naive version and the proper version. We will compare and analyzes the Quality of Result (QoR) differences between them.

Reviewer: ```suggestion
This code sample demonstrates two different implementations of a state machine using SYCL High-Level Synthesis (HLS) on Intel FPGAs: the naive version and the optimal version. We will compare and analyze the Quality of Result (QoR) differences between them.
```

> 

Reviewer: ```suggestion
## Key Implementation Details
```

### Naive Implementation
The naive implementation (`class NaiveStateMachine`) uses basic state machine logic with direct data read and write operations within a loop. Depending on the state, it either processes data in the `State::PROCESS` by multiplying it with a coefficient `coeff` or loads a new coefficient in `State::LD_COEFF` state. State transitions are manually handled within the loop using a swich-case statement, and computation/operations are coded in the body of the each case, then performed sequentially.

> ### Proper Implementation

Reviewer: ```suggestion
### Optimal Implementation
```
i'll stop correcting this :)

The proper implementation (`class ProperStateMachine`) employs the `task_sequence` feature from the SYCL HLS compiler, which allows for more fine-grained control of parallelism and dependency management. Similar to the naive implementation, the state transition and states are defined in a switch-case statement with each case representing a state of the state machine. However, instead of defining the operation inside the case body, this approach defines tasks (`updateCoeff` and `compute`) that are executed asynchronously, enabling the isolation of long initiation intervals (II) and improving throughput by reducing dependency chains between states.

> 

Reviewer: If you weren't already planning to do it, you should include some schematics/code snippets like Juny has in https://github.com/juny-lim/oneAPI-samples-fork/tree/task_seq_hw_reuse/DirectProgramming/C%2B%2BSYCL_FPGA/Tutorials/Features/experimental/hardware_reuse

> ## Limitations of the Naive Version

Reviewer: ```suggestion
### Limitations of the Naive Version
```

The naive implementation suffers from several key limitations:
- **Inefficient State Management:** State transitions and data processing are tightly coupled, leading to increased latency and reduced efficiency in state management.
- **Dependency Bottlenecks:** Each state depends linearly on the completion of the previous state, creating bottlenecks and increasing the total execution time.

![](assets/bottleneck.png)

> ## 3. Improvements Brought by `task_sequence`

Reviewer: ```suggestion
### Improvements brought by `task_sequence`
```

Utilizing `task_sequence` in the proper implementation offers significant enhancements:
- **Enhanced Parallelism:** By decoupling computational dependencies, `task_sequence` allows for more parallel operations, improving overall execution speed and throughput.
- **Reduced Initiation Interval:** The feature isolates unnecessary dependencies between states, which minimizes the initiation interval and maximizes the FPGA's computational efficiency. This enables the compiler to hide the high II from loading the coefficients in the `State::LD_COEFF` state and hence achieves a better overall II for this design.

![](assets/report_screenshot.png)

> 

Reviewer: I'd like you to also include some simulation waveform screenshots to drive home what the impacts look like. Particularly, you should be able to show that the `Optimized` kernel is able to access its `DataIn` and `DataOut` stream every clock cycle, while the `Naive` kernel is not.

Reviewer: ![image](https://github.com/haoyanwa/oneAPI-samples/assets/86071037/32da8700-2443-4bb8-8a89-aca4cee10067)

I tried simulating it, and this is not what we are looking for; i expect there to be **no** gap between the successive pipe reads. Please speak with Johnny (bowen.xue@intel.com) about how to resolve this, he claimed to be able to get II=1.

From the HSD ES Case https://hsdes.intel.com/appstore/article/#/18033853137 

> I did encounter a failure in a different pass when compiling the design (LowerPipes). It was complaining about the datatype of StreamingBeat and expected a struct, so I had to do:
> ```
> struct fake_float {
>  float a;
> };
> using MyStreamingBeat = sycl::ext::intel::experimental::StreamingBeat<fake_float, true, false>;
> ```
> I will be opening a case about this to the memory team.

Use these commands to run the design, depending on your OS.

### On a Linux* System
This design uses CMake to generate a build script for GNU/make.

1. Change to the sample directory.

2. Configure the build system for the Agilex® 7 device family, which is the default.

   ```
   mkdir build
   cd build
   cmake ..
   ```

   You can create a debuggable binary by setting `CMAKE_BUILD_TYPE` to `Debug`:

   ```
   mkdir build
   cd build
   cmake .. -DCMAKE_BUILD_TYPE=Debug
   ```

   If you want to use the `report`, `fpga_sim`, or `fpga` flows, you should switch the `CMAKE_BUILD_TYPE` back to `Release`:

   ```
   cmake .. -DCMAKE_BUILD_TYPE=Release
   ```

   > **Note**: You can change the default target by using the command:
   >  ```
   >  cmake .. -DFPGA_DEVICE=<FPGA device family or FPGA part number>
   >  ```
   >
   > Alternatively, you can target an explicit FPGA board variant and BSP by using the following command:
   >  ```
   >  cmake .. -DFPGA_DEVICE=<board-support-package>:<board-variant>
   >  ```
   >
   > You will only be able to run an executable on the FPGA if you specified a BSP.

> 3. Compile the design with the generated `Makefile`. The following build targets are provided, matching the recommended development flow:
> 
>    | Target          | Expected Time  | Output                                                                       | Description
>    |:---             |:---            |:---                                                                          |:---
>    | `make fpga_emu` | Seconds        | x86-64 binary                                                                | Compiles the FPGA device code to the CPU. Use the Intel® FPGA Emulation Platform for OpenCL™ software to verify your SYCL code’s functional correctness.
>    | `make report`   | Minutes        | RTL + FPGA reports                                                           | Compiles the FPGA device code to RTL and generates an optimization report that describes the structures generated on the FPGA, identifies performance bottlenecks, and estimates resource utilization. This report will include the interfaces defined in your selected Board Support Package. The generated RTL may be exported to Intel® Quartus Prime software.
>    | `make fpga_sim` | Minutes        | RTL + FPGA reports + x86-64 binary                                           | Compiles the FPGA device code to RTL and generates a simulation testbench. Use the Questa*-Intel® FPGA Edition simulator to verify your design.
>    | `make fpga`     | Multiple Hours | Quartus Place & Route (Full accelerator) + FPGA reports + x86-64 host binary | Compiles the FPGA device code to RTL and compiles the generated RTL using Intel® Quartus® Prime. If you specified a BSP with `FPGA_DEVICE`, this will generate an FPGA image that you can run on the corresponding accelerator board.
> 
>    The `fpga_emu`, `fpga_sim` and `fpga` targets produce binaries that you can run. The executables will be called `TARGET_NAME.fpga_emu`, `TARGET_NAME.fpga_sim`, and `TARGET_NAME.fpga`, where `TARGET_NAME` is the value you specify in `CMakeLists.txt`.

Reviewer: ```suggestion
Compile the design. (The provided targets match the recommended development flow.)

Compile for emulation (fast compile time, targets emulates an FPGA device).
make fpga_emu
Generate the HTML optimization reports.
make report
Compile for simulation (fast compile time, targets simulator FPGA device).
make fpga_sim
Compile with Quartus place and route (To get accurate area estimate, longer compile time).
make fpga
```
you can use a simpler text for this (see https://github.com/juny-lim/oneAPI-samples-fork/blob/task_seq_hw_reuse/DirectProgramming/C%2B%2BSYCL_FPGA/Tutorials/Features/experimental/hardware_reuse/README.md)
